src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -475,12 +479,58 @@ ReturnRowsFromTuplestore(uint64 tupleCount, TupleDesc tupleDescriptor,
  * connection so that we receive results a row at a time.
  */
 static bool
-SendQueryInSingleRowMode(PGconn *connection, char *query)
+SendQueryInSingleRowMode(PGconn *connection, char *query, ParamListInfo paramListInfo)
 {
 	int querySent = 0;
 	int singleRowMode = 0;
 
-	querySent = PQsendQuery(connection, query);
+	if (paramListInfo != NULL)
+	{
+		int parameterIndex = 0;
+		int parameterCount = paramListInfo->numParams;
+		Oid *parameterTypes = palloc(parameterCount * sizeof(Oid));
+		const char **parameterValues = palloc(parameterCount * sizeof(char *));
+
+		/* get parameter types and values */
+		for (parameterIndex = 0; parameterIndex < parameterCount; parameterIndex++)
+		{
+			ParamExternData *parameterData = &paramListInfo->params[parameterIndex];
+			Oid typeOutputFunctionId = InvalidOid;
+			bool variableLengthType = false;
+
+			/*
+			 * Use text oid for data types where the oid values can be different
+			 * on the master and worker nodes.
+			 */
+			if (parameterData->ptype >= FirstNormalObjectId)
+			{
+				parameterTypes[parameterIndex] = TEXTOID;
```

Comment:
The reason this fails is that the parameter is unused in your example, because standard_planner() inlined its use. The fix is simple: Murat's router executor change deparses the query using the original Query structure, instead of the modified one. Which means $1 etc. are still used, and thus the types can be inferred.